src/core/thread/time_sync_service.cpp

```diff
@@ -0,0 +1,149 @@
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file implements OpenThread Time Synchronization Service.
+ */
+
+#include "openthread-core-config.h"
+
+#if OPENTHREAD_CONFIG_ENABLE_TIME_SYNC
+
+#include "time_sync_service.hpp"
+
+#include <openthread/platform/alarm-micro.h>
+#include <openthread/platform/alarm-milli.h>
+#include <openthread/platform/time.h>
+
+#include "common/instance.hpp"
+#include "common/logging.hpp"
+
+namespace ot {
+
+TimeSync::TimeSync(Instance &aInstance)
+    : InstanceLocator(aInstance)
+    , mTimeSyncRequired(false)
+    , mTimeSyncSeq(OT_TIME_SYNC_INVALID_SEQ)
+    , mTimeSyncPeriod(OPENTHREAD_CONFIG_TIME_SYNC_PERIOD)
+    , mXtalThreshold(OPENTHREAD_CONFIG_TIME_SYNC_XTAL_THRESHOLD)
+    , mLastTimeSyncSent(0)
+    , mLastTimeSyncReceived(0)
+    , mNetworkTimeOffset(0)
+{
+}
+
+otNetworkTimeStatus TimeSync::GetTime(uint64_t &aNetworkTime) const
+{
+    otNetworkTimeStatus networkTimeStatus;
+
+    if (GetInstance().GetThreadNetif().GetMle().GetRole() <= OT_DEVICE_ROLE_DETACHED)
+    {
+        // The device hasn't attached to a network.
+        networkTimeStatus = OT_NETWORK_TIME_UNSYNCHRONIZED;
+    }
+    else if (GetInstance().GetThreadNetif().GetMle().GetRole() != OT_DEVICE_ROLE_LEADER &&
+             (TimerMilli::GetNow() - mLastTimeSyncReceived) > 2 * TimerMilli::SecToMsec(mTimeSyncPeriod))
+    {
+        // The device hasn’t received time sync for more than two periods time.
+        networkTimeStatus = OT_NETWORK_TIME_RESYNC_NEEDED;
+    }
+    else
+    {
+        networkTimeStatus = OT_NETWORK_TIME_SYNCHRONIZED;
+    }
+
+    aNetworkTime = otPlatTimeGet() + mNetworkTimeOffset;
+
+    return networkTimeStatus;
+}
+
+void TimeSync::HandleTimeSyncMessage(const Message &aMessage)
+{
+    VerifyOrExit(aMessage.GetTimeSyncSeq() != OT_TIME_SYNC_INVALID_SEQ);
```

Comment:
`VerifyOrExit()` here duplicates with the `if` statement when calling `HandleTimeSyncMessage()`?